File: src/Smallest window containing 0, 1 and 2/Code/solution.cpp
class Solution
{
public:
    int smallestSubstring(string s)
    {
        int n = s.size();
        int count[3] = {0, 0, 0};

        int left = 0;
        int ans = INT_MAX;

        for (int right = 0; right < n; right++)
        {
            count[s[right] - '0']++; // include current character

            // Check if window is valid
            while (count[0] > 0 && count[1] > 0 && count[2] > 0)
            {
                ans = min(ans, right - left + 1);

                count[s[left] - '0']--; // remove left character
                left++;                 // shrink window
            }
        }

        return (ans == INT_MAX) ? -1 : ans;
    }
};

File: src/Smallest window containing 0, 1 and 2/Code/solution.py
class Solution:
    def smallestSubstring(self, s):
        n = len(s)
        count = [0, 0, 0]
        
        left = 0
        ans = float('inf')
        
        for right in range(n):
            count[ord(s[right]) - ord('0')] += 1
            
            while count[0] > 0 and count[1] > 0 and count[2] > 0:
                ans = min(ans, right - left + 1)
                
                count[ord(s[left]) - ord('0')] -= 1
                left += 1
        
        return -1 if ans == float('inf') else ans